### PR TITLE
Slack App Integration

### PR DESCRIPTION
**Pull request by sean-brydon**
## What does this PR do?
Slack App setup. 

Implemented the ability for slack to send requests on events (commands etc) - This only works if slacks signature matches with our signature

Fixes # (issue)

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

### Obtaining Slack Client ID and Secret and Signing Secret

To test this you will need to create a Slack app for yourself on [their apps website](https://api.slack.com/apps).

Copy and paste the app manifest below into the setting on your slack app. Be sure to replace `YOUR_DOMAIN` with your own domain or your proxy host if you're testing locally.

<details>
  <summary>App Manifest</summary>

 ```yaml
 display_information:
  name: Cal.com Slack
features:
  bot_user:
    display_name: Cal.com Slack
    always_online: false
  slash_commands:
    - command: /create-event
      url: https://YOUR_DOMAIN/api/integrations/slackmessaging/commandHandler
      description: Create an event within Cal!
      should_escape: false
    - command: /today
      url: https://YOUR_DOMAIN/api/integrations/slackmessaging/commandHandler
      description: View all your bookings for today
      should_escape: false
    - command: /links
      url: https://YOUR_DOMAIN/api/integrations/slackmessaging/commandHandler
      description: View all your Cal.com links
      should_escape: false
oauth_config:
  redirect_urls:
    - https://YOUR_DOMAIN/api/integrations/slackmessaging/callback
  scopes:
    bot:
      - chat:write
      - commands
settings:
  interactivity:
    is_enabled: true
    request_url: https://YOUR_DOMAIN/api/integrations/slackmessaging/interactiveHandler
    message_menu_options_url: https://YOUR_DOMAIN/api/integrations/slackmessaging/interactiveHandler
  org_deploy_enabled: false
  socket_mode_enabled: false
  token_rotation_enabled: false
```

</details>

Add the integration as normal - slack app - add. Follow the oauth flow to add it to a server.

Next make sure you have your app running `yarn dx`. Then in the slack chat type one of these commands: `/create-event` or `/today` or `/links`

> NOTE: Next you will need to setup a proxy server like [ngrok](https://ngrok.com/) to allow your local host machine to be hosted on a public https server.

## Checklist

- I haven't commented my code, particularly in hard-to-understand areas
- I haven't checked if my PR needs changes to the documentation
- I haven't added tests that prove my fix is effective or that my feature works

### CODE REVIEW

File: apps/web/next.config.js [hunk 1]
```diff
@@ -1,4 +1,4 @@
-const withTM = require("@vercel/edge-functions-ui/transpile")([
+const withTM = require("next-transpile-modules")([
```
Comment (zomars) [inline apps/web/next.config.js:1]: Can we separate this into a another PR directly to `main`?

Comment (zomars) [reply apps/web/next.config.js:1]: Don't need to remove them, once this lands on main this diff will be gone.

File: apps/web/pages/api/integrations/slackapp/listener/[...args].ts [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+    return res.status(400).json({ message: "Request is too old" });
+  }
+
+  const hmac = createHmac("sha256", signingSecret);
+  const [version, hash] = slackSignature.split("=");
+
+  const payload = body.payload && JSON.parse(body.payload);
+  console.log(payload);
+
+  hmac.update(`${version}:${timeStamp}:${stringify(payload ? { payload } : body)}`);
```
Comment (sean-brydon) [inline apps/web/pages/api/integrations/slackapp/listener/[...args].ts:39]: See: https://api.slack.com/authentication/verifying-requests-from-slack for the hashing side of this verification

File: apps/web/pages/api/integrations/slackapp/listener/[...args].ts [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+  const digest = hmac.digest("hex");
+
+  // TODO: This still fails on and custom selections - need to figure out how to handle this. Normal slack actions do not produce an invalid signature.
+  if (digest !== hash) {
+    return res.status(400).json({ message: "Invalid signature" });
+  }
+}
+
+export default async function handler(req: NextApiRequest, res: NextApiResponse) {
```
Comment (sean-brydon) [inline apps/web/pages/api/integrations/slackapp/listener/[...args].ts:48]: A new handler was created to handle slack auth requests and that is it - this is because the normal [...args] route that we have defined will automatically fail on verification

File: packages/app-store/slackapp/api/add.ts [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+      select: {
+        id: true,
+      },
+    });
+    const params = {
+      client_id,
+      scope: scopes.join(","),
+    };
+    const query = stringify(params);
+    const url = `https://slack.com/oauth/v2/authorize?${query}&user_`;
```
Comment (sean-brydon) [inline packages/app-store/slackapp/api/add.ts:26]: Generate the oauth URL

File: packages/app-store/slackapp/api/callback.ts [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+
+    await prisma.user.update({
+      where: {
+        id: req.session?.user.id,
+      },
+      data: {
+        credentials: {
+          create: {
+            type: "slack_app",
+            key: responseBody,
```
Comment (sean-brydon) [inline packages/app-store/slackapp/api/callback.ts:39]: Add the slack credetials to the users object

File: packages/app-store/slackapp/api/commandHandler.ts [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+import prisma from "@calcom/prisma";
+
+import { showCreateEventMessage, showTodayMessage } from "../lib";
+
+export enum SlackAppCommands {
+  CREATE_EVENT = "create-event",
+  TODAY = "today",
+}
+
+export default async function handler(req: NextApiRequest, res: NextApiResponse) {
```
Comment (sean-brydon) [inline packages/app-store/slackapp/api/commandHandler.ts:12]: Easier way to handle commands instead of creating a new endpoint per command

File: packages/app-store/slackapp/api/interactiveHandler.ts [hunk 7]
```diff
@@ -0,0 +1,24 @@
+import { NextApiRequest, NextApiResponse } from "next";
+import { parse } from "querystring";
+
+import createEvent from "../lib/actions/createEvent";
+
+enum InteractionEvents {
+  CREATE_EVENT = "cal.event.create",
+}
+
+export default async function interactiveHandler(req: NextApiRequest, res: NextApiResponse) {
```
Comment (sean-brydon) [inline packages/app-store/slackapp/api/interactiveHandler.ts:10]: Handle interactions - any button clicks in slack that require processing to happen

File: packages/app-store/slackapp/lib/actions/createEvent.ts [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+export default async function createEvent(req: NextApiRequest, res: NextApiResponse) {
+  const {
+    user,
+    view: {
+      state: { values },
+    },
+  } = JSON.parse(req.body.payload);
+
+  // This is a mess I have no idea why slack makes getting infomation this hard.
+  const {
```
Comment (sean-brydon) [inline packages/app-store/slackapp/lib/actions/createEvent.ts:44]: Grab users input from the slack form

File: packages/app-store/slackapp/lib/actions/createEvent.ts [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+  // This could get a bit weird as there is a 3 second limit until the post times ou
+
+  // Compute all users that have been selected and get their email.
+  const invitedGuestsEmails = selected_users.map(
+    async (userId: string) => await getUserEmail(client, userId)
+  );
+
+  const startDate = dayjs(`${selected_date} ${selected_time}`, "YYYY-MM-DD HH:mm");
+
+  const PostData: BookingCreateBody = {
```
Comment (sean-brydon) [inline packages/app-store/slackapp/lib/actions/createEvent.ts:107]: Create the booking object to create the booking for the user

File: packages/app-store/slackapp/lib/actions/createEvent.ts [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+    start: dayjs(startDate).format(),
+    end: dayjs(startDate)
+      .add(foundUser?.eventTypes[0]?.length ?? 0, "minute")
+      .format(),
+    eventTypeId: foundUser?.eventTypes[0]?.id ?? 0,
+    user: foundUser?.username ?? "",
+    email: foundUser?.email ?? "",
+    name: selected_name,
+    guests: await Promise.all(invitedGuestsEmails),
```
Comment (sean-brydon) [inline packages/app-store/slackapp/lib/actions/createEvent.ts:116]: Transform all selected users in the slack selection to email address

File: packages/app-store/slackapp/lib/showCreateEventMessage.ts [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+      },
+    },
+  });
+
+  if (!data) return res.status(200).json(NoUserMessage);
+  const slackCredentials = data?.key; // Only one slack credential for user
+  // @ts-ignore access_token must exist on slackCredentials otherwise we have wouldnt have reached this endpoint
+  const access_token = slackCredentials?.access_token;
+  const slackClient = new WebClient(access_token);
+  await slackClient.views.open({
```
Comment (sean-brydon) [inline packages/app-store/slackapp/lib/showCreateEventMessage.ts:32]: Open the create event modal

File: packages/app-store/slackapp/lib/showTodayMessage.ts [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+          id: true,
+          email: true,
+        },
+      },
+    },
+  });
+
+  if (!foundUser) res.status(200).json(NoUserMessage);
+
+  const bookings = await prisma.booking.findMany({
```
Comment (sean-brydon) [inline packages/app-store/slackapp/lib/showTodayMessage.ts:25]: Get any bookings for that current day

File: packages/app-store/slackapp/lib/showTodayMessage.ts [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+            some: {
+              email: foundUser?.user?.email,
+            },
+          },
+        },
+      ],
+      AND: [
+        {
+          endTime: { gte: new Date() },
+          AND: [
```
Comment (joeauyeung) [inline packages/app-store/slackapp/lib/showTodayMessage.ts:42]: Do we include bookings that require opt in and `confirmed: false`?

Comment (sean-brydon) [reply packages/app-store/slackapp/lib/showTodayMessage.ts:42]: Currently no - as from what i can tell we are fetching the data the exact same way as bookings/upcoming

File: packages/app-store/slackapp/api/commandHandler.ts [hunk 14]
```diff
@@ -0,0 +1,26 @@
...
+  if (req.method === "POST") {
+    const command = req.body.command.split("/").pop();
+
+    switch (command) {
+      case SlackAppCommands.CREATE_EVENT:
```
Comment (alannnc) [inline packages/app-store/slackapp/api/commandHandler.ts:17]: probably a fetch from user availability here? to remove already booked days/time slots

Comment (sean-brydon) [reply packages/app-store/slackapp/api/commandHandler.ts:17]: The issue is - the datapicker in slack doesn't allow you to hide times / dates from what I have found

File: apps/web/pages/api/integrations/[...args].ts [hunk 15]
```diff
@@ -9,11 +9,6 @@ const handler = async (req: NextApiRequest, res: NextApiResponse) => {
   // Check that user is authenticated
   req.session = await getSession({ req });
 
-  if (!req.session?.user?.id) {
-    res.status(401).json({ message: "You must be logged in to do this" });
-    return;
-  }
```
Comment (zomars) [inline apps/web/pages/api/integrations/[...args].ts:15]: We'll let each app to decide if they need authed handlers

File: packages/app-store/components.tsx [hunk 16]
```diff
@@ -14,6 +14,7 @@ export const InstallAppButtonMap = {
   caldavcalendar: dynamic(() => import("./caldavcalendar/components/InstallAppButton")),
   googlecalendar: dynamic(() => import("./googlecalendar/components/InstallAppButton")),
   office365calendar: dynamic(() => import("./office365calendar/components/InstallAppButton")),
+  slackapp: dynamic(() => import("./slackapp/components/InstallAppButton")),
```
Comment (zomars) [inline packages/app-store/components.tsx]: Adds missing install button

File: packages/prisma/schema.prisma [hunk 17]
```diff
@@ -7,7 +7,8 @@ datasource db {
 }
 
 generator client {
-  provider = "prisma-client-js"
+  provider        = "prisma-client-js"
+  previewFeatures = ["filterJson"]
```
Comment (zomars) [inline packages/prisma/schema.prisma:11]: Did we end up using this? If not lets remove it.

**Review by zomars — COMMENTED** [review]
We're almost there. Let's just figure out the connection error on the Slack modal.

**Review by zomars — APPROVED** [review]

**Review by zomars — CHANGES_REQUESTED** [review]
Let's solve these type errors @sean-brydon 

image

**Review by zomars — APPROVED** [review]
